File: WebAPI开发手册/demo/cloud_websocket_tts_test.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

"""
python 3.7

# websocket-client版本要求不低于 0.57.0
pip install websocket-client==0.57.0
"""

import websocket
from email.utils import formatdate
import hmac
from hashlib import sha256
import base64
import json
from urllib.parse import urlencode
import _thread as thread

# 由北京市大牛儿科技发展有限公司统一分配
org_id = 0
app_id = 0
app_key = ""

# 请求的地址
request_url = "ws://api.baller-tech.com/v1/service/ws/v1/tts"
host = "api.baller-tech.com"

# 测试使用的文本数据
txt_file = ""
# 测试的语种
language = ""
# 合成的音频文件格式
# 请查考《语音识别（TTS）HTTP协议WebAPI开发文档.pdf》中“支持的语种以及采样格式”章节
sample_format = "audio/L16;rate=16000"
# 合成的音频的压缩类型
# 请查考《语音识别（TTS）HTTP协议WebAPI开发文档.pdf》中“支持的音频编码”章节
audio_encode = "raw"


def on_error(ws, error):
    print(f"ERROR: {error}")


# 一次语音合成任务只需发送一次文本数据，但合成的结果会分多次返回。
# 服务端会将传入的文本分为不同的子句，每次on_message触发时，返回的是一个子句的合成结果，当收到一个子句的合成结果时，应用就可以开始播放，下一个子句的合成结果会在当前子句播放完成前返回。
def on_message(ws, message):
    try:
        message_values = json.loads(message)
    except Exception as e:
        print(f"load message {message} failed {e}")
        return

    # task_id只在服务器推送的第一帧中出现
    if "task_id" in message_values:
        ws.result_file = open(f"{message_values['task_id']}.{audio_encode}", "wb")
        print(f"task id {message_values['task_id']}")

    if 0 != message_values["code"]:
        print(f"tts failed {message_values['code']} {message_values['message']}")
        return

    # 将合成结果保存到文件
    if message_values["data"]:
        if ws.result_file:
            ws.result_file.write(base64.b64decode(message_values['data']))

    # 最后一帧时关闭WebSocket连接
    if 1 == message_values["is_end"]:
        ws.close()


def on_open(ws):
    def run(**kwargs):
        with open(txt_file, "rb") as fp:
            txt_data = fp.read()
            fp.close()

        # 业务参数
        business_params = {
            "language": language,
            "sample_format": sample_format,
            "audio_encode": audio_encode,
        }
        data_params = {
            "txt": base64.b64encode(txt_data).decode(encoding='utf-8'),
        }
        params = {"data": data_params, "business": business_params}
        ws.send(json.dumps(params))

    thread.start_new_thread(run, ())


def test_ws():
    date = formatdate(timeval=None, localtime=False, usegmt=True)

    # signature
    signature_org = ""
    signature_org += "app_id:" + str(app_id) + "\n"
    signature_org += "date:" + date + "\n"
    signature_org += "host:" + host
    signature_sha = hmac.new(app_key.encode(encoding='utf-8'),
                             signature_org.encode(encoding='utf-8'),
                             digestmod=sha256
                             ).digest()
    signature_final = base64.b64encode(signature_sha).decode(encoding='utf-8')

    # authorization
    authorization_org = {
        "app_id": str(app_id),
        "signature": signature_final,
    }
    authorization_org = json.dumps(authorization_org).encode(encoding='utf-8')
    authorization_final = base64.b64encode(authorization_org).decode(encoding='utf-8')

    url_params = {
        "date": date,
        "host": host,
        "authorization": authorization_final
    }
    url_with_param = request_url + "?" + urlencode(url_params)

    websocket.enableTrace(False)
    ws = websocket.WebSocketApp(url_with_param,
                                on_message=on_message,
                                on_error=on_error,
                                on_open=on_open)

    ws.run_forever()


if __name__ == '__main__':
    test_ws()

File: WebAPI开发手册/demo/cloud_websocket_tts_test.java
package com.baller;

import java.io.BufferedInputStream;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URLEncoder;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.text.SimpleDateFormat;
import java.util.Base64;
import java.util.Calendar;
import java.util.Locale;
import java.util.TimeZone;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft_6455;
import org.java_websocket.handshake.ServerHandshake;

import com.alibaba.fastjson.JSONException;
import com.alibaba.fastjson.JSONObject;

/**
 * 
 * 信息配置类
 */
class Common {
    // 地址配置
    public static String mUrl = "ws://api.baller-tech.com/v1/service/ws/v1/tts";
    public static String mHost = "api.baller-tech.com";

    // 账号信息 由北京市大牛儿科技发展有限公司统一分配
    public static long mOrgId = 0L;
    public static long mAppId = 0L;
    public static String mAppkey = "";

    // 语种
    // 请查考《语音识别（TTS）WebSocket协议WebAPI开发文档.pdf》中“支持的语种以及采样格式”章节
    public static String mLanguage = "";
    // 测试使用的文本文件
    public static String mTxtFile = "";
    // 合成的音频文件格式
    // 请查考《语音识别（TTS）WebSocket协议WebAPI开发文档.pdf》中“支持的语种以及采样格式”章节
    public static String mSampleFormat = "audio/L16;rate=16000";
	// 合成语速
	// 请查考《语音识别（TTS）WebSocket协议WebAPI开发文档.pdf》中“语速的取值范围”章节
	public static float mSpeed = 1.0f;
    // 合成的音频的压缩类型，
    // 请查考《语音识别（TTS）WebSocket协议WebAPI开发文档.pdf》中“支持的音频编码”章节
    public static String mAudioEncode = "raw";
    // 结果保存文件
    public static String mSaveResultFile = mTxtFile + "_out." + mAudioEncode;

    // 读取文件
    @SuppressWarnings("resource")
    public static byte[] readFile(String strFile) {
        byte[] fileData = null;

        BufferedInputStream inputStream;
        try {
            inputStream = new BufferedInputStream(new FileInputStream(strFile));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            return fileData;
        }

        try {
            fileData = new byte[inputStream.available()];
        } catch (IOException e) {
            e.printStackTrace();
            return fileData;
        }

        try {
            inputStream.read(fileData);
        } catch (IOException e) {
            e.printStackTrace();
            return fileData;
        }

        return fileData;
    }
}

/**
 * 
 * 发送数据的线程类
 */
class SendFrameThread extends Thread {
    private WebSocketClient mClient;
    private boolean mIsFirstFrame = true;

    SendFrameThread(WebSocketClient wsClient) {
        this.mClient = wsClient;
    }

    public String makeFrame(byte[] data) {
        JSONObject jsonParams = new JSONObject();
        if (mIsFirstFrame) {
            JSONObject jsonBusiness = new JSONObject();
            try {
                jsonBusiness.put("language", Common.mLanguage);
                jsonBusiness.put("sample_format", Common.mSampleFormat);
                jsonBusiness.put("audio_encode", Common.mAudioEncode);
				jsonBusiness.put("speed", Common.mSpeed);
                jsonParams.put("business", jsonBusiness);
            } catch (JSONException e) {
                e.printStackTrace();
                return null;
            }

            mIsFirstFrame = false;
        }

        JSONObject jsonData = new JSONObject();
        try {
            jsonData.put("txt", Base64.getEncoder().encodeToString(data));
            jsonParams.put("data", jsonData);
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }

        return jsonParams.toString();
    }

    public void run() {
        // 读取测试文件
        byte[] imageData = Common.readFile(Common.mTxtFile);
        if (imageData == null || imageData.length == 0) {
            System.out.println("读取测试文件出错");
            return;
        }

        String strSendFrame = this.makeFrame(imageData);
        if (strSendFrame != null) {
            if (mClient.isClosed() || mClient.isClosing()) {
                return;
            }
            mClient.send(strSendFrame);
        } else {
            return;
        }
    }
}

public class cloud_websocket_tts_test {

    private static DataOutputStream mOutFile = null;

    private static String HMACSHA256AndBase64(byte[] data, byte[] key) {
        try {
            SecretKeySpec signingKey = new SecretKeySpec(key, "HmacSHA256");
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(signingKey);
            return Base64.getEncoder().encodeToString(mac.doFinal(data));
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (InvalidKeyException e) {
            e.printStackTrace();
        }

        return null;
    }

    private static String makeAuthorization() {
        Calendar cd = Calendar.getInstance();
        SimpleDateFormat sdf = new SimpleDateFormat("EEE d MMM yyyy HH:mm:ss 'GMT'", Locale.US);
        sdf.setTimeZone(TimeZone.getTimeZone("GMT"));
        String strRequestDateTime = sdf.format(cd.getTime());

        String strSignatureOrg = "app_id:" + Common.mAppId + "\n";
        strSignatureOrg += "date:" + strRequestDateTime + "\n";
        strSignatureOrg += "host:" + Common.mHost;
        String strSignatureFinal = HMACSHA256AndBase64(strSignatureOrg.getBytes(), Common.mAppkey.getBytes());

        JSONObject jsonAuthorization = new JSONObject();
        try {
            jsonAuthorization.put("app_id", Long.toString(Common.mAppId));
            jsonAuthorization.put("signature", strSignatureFinal);
        } catch (JSONException e) {
            e.printStackTrace();
            return "";
        }

        String strAuthorizationBase64 = Base64.getEncoder().encodeToString(jsonAuthorization.toString().getBytes());
        String strAuthorizationFinal = "";
        try {
            strAuthorizationFinal = "authorization=" + URLEncoder.encode(strAuthorizationBase64, "UTF-8") + "&host="
                    + URLEncoder.encode(Common.mHost, "UTF-8") + "&date="
                    + URLEncoder.encode(strRequestDateTime, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

        return strAuthorizationFinal;
    }

    public static void main(String[] args) {
        // 检查运行条件
        if (Common.mAppId == 0L || Common.mAppkey == "") {
            System.out.println("请填写账号信息");
            return;
        }

        try {
            mOutFile = new DataOutputStream(new FileOutputStream(Common.mSaveResultFile, true));
        } catch (FileNotFoundException e1) {
            System.out.println("打开结果文件" + Common.mSaveResultFile + "失败");
            e1.printStackTrace();
            return;
        }

        // 开始连接
        WebSocketClient wsClient = null;
        try {
            String strUrl = Common.mUrl + "?" + makeAuthorization();
            wsClient = new WebSocketClient(new URI(strUrl), new Draft_6455()) {
                @Override
                public void onOpen(ServerHandshake handshakedata) {
                    // 启动数据发送线程
                    SendFrameThread sendThread = new SendFrameThread(this);
                    sendThread.start();
                }

                // 一次语音合成任务只需发送一次文本数据，但合成的结果会分多次返回。
                // 服务端会将传入的文本分为不同的子句，每次onMessage触发时，返回的是一个子句的合成结果，当收到一个子句的合成结果时，应用就可以开始播放，下一个子句的合成结果会在当前子句播放完成前返回。
                @Override
                public void onMessage(String message) {
                    JSONObject jsonObject = JSONObject.parseObject(message);
                    int iCode = jsonObject.getInteger("code");

                    if (0 == iCode) {
                        // 正常流程
                        boolean isEnd = jsonObject.getBoolean("is_end");
                        String strDataAfterBase64 = jsonObject.getString("data");
                        byte[] arData = Base64.getDecoder().decode(strDataAfterBase64);

                        try {
                            mOutFile.write(arData);
                        } catch (IOException e) {
                            e.printStackTrace();
                        }

                        if (isEnd) {
                            // 处理完毕主动关闭连接
                            this.close();
                            System.out.println("语音合成完毕");
                        }
                    } else {
                        // 出错流程
                        String strMessage = jsonObject.getString("message");
                        System.out.println("服务器报错 code: " + iCode + " message: " + strMessage);
                        this.close();
                    }
                }

                @Override
                public void onClose(int code, String reason, boolean remote) {
                    if (reason.length() > 0) {
                        System.out.println("[websocket] 关闭(" + reason + ")");
                    }
                }

                @Override
                public void onError(Exception e) {
                    System.out.println("[websocket] 出错 ");
                    e.printStackTrace();
                }
            };
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        // 开始建立连接
        if (wsClient != null) {
            wsClient.connect();
        }
    }
}
